### PR TITLE
WIP: Adding TimeAdjustment code.  Takes MSID and returns a modified MSID.

### PR DESCRIPTION
**Pull request by jmodelcxc**
…ith timestamps corrected for START_MINOR_FRAME on a per-sample basis

## Testing

- [ ] Passes unit tests on MacOS, linux, Windows (at least one required)
- [ ] Functional testing

Fixes #

### CODE REVIEW

**Review by jmodelcxc — COMMENTED** [review]
Looks good.  Thanks for tidying.

**Comment by taldcroft** [comment]
@jmodelcxc - thanks, this looks like a great start. I just skimmed the actual code and see that it should be fine for inclusion.

Eventually this will be a function that typically just gets called within `fetch.MSID`, but the way you've written it now is good.

First order of business then is to make the code compliant with the [Ska Python coding standards](https://occweb.cfa.harvard.edu/twiki/Aspect/SkaPython). For better or worse, we are fairly strict on following industry best-practices for code formatting, variable naming, documentation style conventions etc as defined in PEP8. This is pretty important for a large production code base. The first thing is that your code should be clean when you run this:
```
flake8 time_adjust.py --ignore=E402,W503,F541 --max-line-length=100
```

In order to get you on the right track, I ran `autopep8` to fix most of the issues, and then did some edits myself to clean up much of the rest.

The next thing you can do is move the tests that are currently in `main()` into proper unit tests in `Ska/eng_archive/tests/test_fetch.py`. After that you can test with something like:
```
python setup.py test --args='-k test_time_adjust'
```

File: Ska/engarchive/time_adjust.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Describe what this module does.
+"""
+
+import copy
```
Comment (taldcroft) [inline Ska/engarchive/time_adjust.py:6]: Imports ordered in sections: standard library imports then other installed packages.

File: Ska/engarchive/time_adjust.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    - TBD: add an option for in-place adjustment
+    - TBD: - maybe take msid as string, but manually set data source to avoid
+      'maude' 'cxc' differences...
+    - NOTE: Could use times from the msid.times span, but may create problems at
+      the beginning and end since these are subject to adjustment.  Should use
+      time query that generated the MSID
+
+    :param msid: fetch ``MSID`` or ``Msid`` object
+        MSID for which to calculate the time adjustments statistics for.
+    :param ts: str, ``CxoTime``, ``DateTime``
```
Comment (taldcroft) [inline Ska/engarchive/time_adjust.py:34]: Standard naming for generic start times is `start`, and likewise `stop`. If a time is specifically a float in CXCtime then `tstart`  / `tstop`.

**Comment by jmodelcxc** [comment]
Will do on the coding style.  I'm fighting decades of individual coder bad-habits.  


> The next thing you can do is move the tests that are currently in `main()` into proper unit tests in `Ska/eng_archive/tests/test_fetch.py`. After that you can test with something like:
> 
> ```
> python setup.py test --args='-k test_time_adjust'
> ```
OK, that looks straightforward.  Just assert a passing test condition.

File: Ska/engarchive/time_adjust.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    # Generate list of intervals for each format using logical intervals
+    fmts = ('FMT1', 'FMT2', 'FMT3', 'FMT4', 'FMT5', 'FMT6')
+    tmf_intervals = {}
+    for fmt in fmts:
+        tmf_intervals[fmt] = logical_intervals(tmf.times, tmf.vals == fmt, complete_intervals=False)
+
+    # Make scratchpad copy to avoid in-place effects (I think need to check by
+    # ref or by val convention)
+    times = msid.times
```
[the remark this answers is not in the document]
Comment (jmodelcxc) [reply Ska/engarchive/time_adjust.py:84]: Looking at the code more closely, the condition is checked for the entire vector BEFORE modifications are made.  So just removing the 'times' variable entirely and self incrementing should be fine.  

> msid.times[(msid.times >= interval['tstart']) & (msid.times < interval['tstop'])
                  ] += t_off[fmt]

**Comment by jeanconn** [comment]
I haven't examined the history in detail but our more common practice is to rebase from main/master instead of merging.  Another option is to just rebase the PR when done to squash commits.

And still not sure if we need to remake this PR due to the odd "build" error.

**Comment by jmodelcxc** [comment]
> I haven't examined the history in detail but our more common practice is to rebase from main/master instead of merging. Another option is to just rebase the PR when done to squash commits.
> 
> And still not sure if we need to remake this PR due to the odd "build" error.
Hi Jean, 
What was the "odd build error"?   When that's all resolved  I understand your second technique (i.e.  rebase the PR to squash the commits) but I don't quite get how rebase from main/master would work. 
Thanks

**Comment by jeanconn** [comment]
With regard to "I don't quite get how rebase from main/master would work.", for a lot of our work, if we're developing in a branch (that started from main) and the branch has commits 1, 2... and in the meantime unrelated commits 3, 4 have ended up in main, if we "git rebase -i origin/main" in the branch, most of the time it will rebase cleanly and the branch history will now have 3, 4, 1, 2 and the branch will only differ from main after commit 4.  

But I may be using the wrong language... and there are certainly tons of git branch merge/rebase diagrams and tutorials out there.

**Comment by jeanconn** [comment]
With regard to "odd build error", this PR is still stuck at the bottom with "Some checks haven't completed yet" on a "build" that looks stuck.

**Comment by jmodelcxc** [comment]
> With regard to "odd build error", this PR is still stuck at the bottom with "Some checks haven't completed yet" on a "build" that looks stuck.
Ah - are those checks run automatically (some kind of CI tool?)  If so is there place to see the logs?  I didn't manually "build" anything, just added/committed my files to my branch.

**Comment by jmodelcxc** [comment]
> With regard to "I don't quite get how rebase from main/master would work.", for a lot of our work, if we're developing in a branch (that started from main) and the branch has commits 1, 2... and in the meantime unrelated commits 3, 4 have ended up in main, if we "git rebase -i origin/main" in the branch, most of the time it will rebase cleanly and the branch history will now have 3, 4, 1, 2 and the branch will only differ from main after commit 4.
> 
> But I may be using the wrong language... and there are certainly tons of git branch merge/rebase diagrams and tutorials out there.


Got it (I think), that makes sense now.  So I wouldn't need to make a new PR for that, just keep my branch abreast of the changing master

**Comment by jeanconn** [comment]
Exactly.  Those are automatic checks, and we *think* it should be running https://github.com/sot/eng_archive/blob/master/.github/workflows/flake8.yml .  But unfortunately, it looks like github isn't giving us *anything* for this weird "Waiting for status to be reported" state.  We generally get complete logs for something like a real failure in flake8 (we just saw one while working in cxotime) https://github.com/sot/cxotime/runs/1856064359 .

And I can't find anything in the github help for dealing with a "stuck" automatic check , so was somewhat wondering about making a new PR just to try to get around that issue.

Generally, we update (rebase as needed, review, add new related commits) all in the one PR that is on one conceptual change in the code (as much as that makes sense).  We often don't even need to rebase because the changes in main may not have anything to do with review of the PR... which is why I started with "I haven't examined the history in detail".  I just saw your commit messages with merges from main/master and wondered.

Sometimes, per my previous numbered example, it is necessary to get commits 3, 4 into the branch history while working on the commit 1, 2 changes... if something in 3, 4 *is* actually related, or just needed to test appropriately etc.

**Comment by taldcroft** [comment]
Not likely to get merged. It'll be here if it ever comes up again.

**Pull request closed without merging**